zospy/analyses/polarization.py

```diff
@@ -0,0 +1,344 @@
+    cfgoutfile: str or None
+        The configuration file to which the current configuration is saved. If None, a temporary file will be created
+        and deleted. If string, it should be a full system path with '.CFG' as extension, after which the file will be
+        saved in that location. Defaults to
+        None.
+    txtoutfile: str or None
+        The textfile to which the OpticStudio output is saved. If None, a temporary file will be created and deleted. If
+        string, it should be a full system path with '.txt' as extension, after which the file will be saved in that
+        location. Defaults to None.
+
+    Returns
+    -------
+    AnalysisResult
+        A Polarization Pupil Map. Next to the standard data, the raw text return obtained from the analysis
+        will be present under 'RawTextData', and the txtoutfile under 'TxtOutFile'.
+    """
+
+    analysistype = 'PolarizationPupilMap'
+
+    if cfgoutfile is None:
+        fd, cfgoutfile = mkstemp(suffix='.CFG', prefix='zospy_')
+        os.close(fd)
+        cleancfg = True
+    else:
+        if not cfgoutfile.endswith('.CFG'):
+            raise ValueError('cfgfile should end with ".CFG"')
+        cleancfg = False
+
+    if txtoutfile is None:
+        fd, txtoutfile = mkstemp(suffix='.txt', prefix='zospy_')
+        os.close(fd)
+        cleantxt = True
+    else:
+        if not txtoutfile.endswith('.txt'):
+            raise ValueError('txtfile should end with ".txt"')
+        cleantxt = False
+
+    analysis = oss.Analyses.New_Analysis_SettingsFirst(constants.Analysis.AnalysisIDM.loc[analysistype])
+
+    # Modify the settings file
+    an_sett = analysis.GetSettings()
+    an_sett.SaveTo(cfgoutfile)
+
+    # MODIFYSETTINGS are defined in the ZPL help files: The Programming Tab > About the ZPL > Keywords
+    an_sett.ModifySettings(cfgoutfile, 'PPM_JX', str(jx))
+    an_sett.ModifySettings(cfgoutfile, 'PPM_JY', str(jy))
+    an_sett.ModifySettings(cfgoutfile, 'PPM_PX', str(x_phase))
+    an_sett.ModifySettings(cfgoutfile, 'PPM_PY', str(y_phase))
+    an_sett.ModifySettings(cfgoutfile, 'PPM_WAVE', str(int(wavelength)))
+    an_sett.ModifySettings(cfgoutfile, 'PPM_FIELD', str(int(field)))
+    an_sett.ModifySettings(cfgoutfile, 'PPM_SURFACE', str(surface))
+    samp = utils.zputils.proc_constant(constants.Analysis.SampleSizes_ContrastLoss,
+                                 utils.zputils.standardize_sampling(sampling))
+    an_sett.ModifySettings(cfgoutfile, 'PPM_SAMP', str(samp-1))
+    an_sett.ModifySettings(cfgoutfile, 'PPM_ADDCONFIG', str(add_configs))
+    an_sett.ModifySettings(cfgoutfile, 'PPM_SUBCONFIGS', str(sub_configs))
+
+    an_sett.LoadFrom(cfgoutfile)
+
+    # Run analysis
+    analysis.ApplyAndWaitForCompletion()
+
+    # Get results
+    analysis.Results.GetTextFile(txtoutfile)
+    line_list = [line for line in open(txtoutfile, 'r', encoding='utf-16-le')]
+
+    # Create output dict
+    data = AttrDict()

```

Comment:
Could you give me an example on how you implemented dataclasses instead of AttrDict? I could not find any in the other files. It seems that in some cases AttrDict is still used.

Reply:
You're right, we'll convert the existing analyses later on. The idea is quite simple: define a dataclass with the same attributes that are used in the `AttrDict`. This may be less "smart" than the `AttrDict`, but at least you know what you get without having to interactively inspect the results, and [`AttrDict`s can cause some nasty problems](https://stackoverflow.com/questions/4984647/accessing-dict-keys-like-an-attribute).